032-projects/proposal.md

```diff
@@ -0,0 +1,274 @@
+# Summary
+
+A Concourse project is a namespace for resources, tasks, and pipelines. Projects are bootstrapped from a single resource definition and a build plan which executes when new versions of the resource are found. Projects bridge the gap between small-scale and large-scale CI workflows, helping to smoothen Concourse's learning curve.
+
+# Motivation
+
+* Smoothen the learning curve. Right now the jump from `fly execute` to `fly set-pipeline` is pretty steep. Projects can be a user's first introduction to build plans, which transitions nicely into jobs and pipelines.
+* Enable small-scale projects to have a simple Travis/Circle CI single-build workflow, without needing pipelines at all. This is a scenario where Concourse pipelines can be seen as overkill.
+* Enable medium-to-large-scale projects to fully automate their CI stack by using the project's build plan to configure pipelines via the [`set_pipeline` step](https://github.com/concourse/rfcs/pull/31).
+* Enable cross-pipeline `passed` constraints between pipelines within the project. This will allow large pipelines to be broken up into reusable templates while still supporting fan-in downstream.
+* Ensure tasks and pipelines are always in sync by having both of them come from the project instead of passing task configs as a resource throughout the pipeline.
+
+# Proposal
+
+A Concourse project allows you to bootstrap and automate your CI/CD stack by configuring a single resource which contains your project's config:
+
+```sh
+fly set-project \
```

Comment:
Maybe this is a bit too much scope-creep, but I would propose including some mechanism for bootstrapping Concourse with a list of pre-defined projects - some kind of `initial_config.yaml` that includes teams and projects belonging to those teams, that can be checked in with whatever is used to setup Concourse in the first place.

As somebody who loves Concourse's strongly opinionated commitment to reducing state - [`worker-state`](https://github.com/concourse/concourse/wiki/Anti-Patterns#worker-state) being a recognized anti-pattern - it would be nice to see a nod to trying to reduce `web-state` or `atc-state`. One of the advantages of using the pipeline resource over `fly set-pipeline` was having one single `meta` pipeline, checked into version control, which bootstrapped the rest of the pipelines. Is there any way here to have one single thing of some kind which can bootstrap the rest of the projects on the `atc`?

Reply:
@pivotal-jwinters, @pivotal-jamie-klassen and I chatted about this as sort of the natural conclusion to this trend. I think it's a great idea but we should get there in the future as a separate RFC. It doesn't replace 'projects' since as you pull on the thread the required role of the user making changes goes from 'team member' to 'team admin' to 'concourse operator'.

So much in the same way that Pipelines were manually configured until Projects were introduced to bootstrap them, I imagine Projects will be manually configured until we introduce something higher up. Eventually we'll end up with all the `web` config coming from a `git` repo, I imagine. :stuck_out_tongue:

Reply:
@vito I think that would be a great idea.

On the required role of operator, I'd argue that there's two perspectives:
only requiring team members is great for flexibility, and flexibility is
certainly something that Concourse has excelled at - i.e. `fly execute`. On
the other hand, if the flip side of requiring the cluster operator is that
projects, pipelines, etc. simply *exist*, completely automatically
according to templates, including required verification, notification, etc.
steps, having been setup automatically when new repositories were created,
then, the flexibility just isn't needed at all. I think Concourse can lend
itself very well to the second model as well - usable either way, depending
on the culture and needs of the group running the Concourse instance.

Reply:
Right. What you're describing sounds like a single-tenant use case, where everyone involved in the Concourse installation is aware of how it's all set up and has a workflow that encompasses the whole Concourse cluster. This is true of a lot of (maybe most) Concourse installations.

So far, I'm approaching this from a multi-tenant viewpoint where things are more 'self-service' and users who may not know each other are sharing a Concourse and need to be empowered to configure their own things without having admin capabilities or a central authority merging pull requests in a Concourse cluster automation repo. I think this will be easier to work with if/when we ever want to do a hosted Concourse offering.

Side tangent: back in the day, Concourse was literally configured with a single pipeline. It wasn't even called a pipeline, it was just a config file. I resisted going any further than that for a while, wishing people would just run a different Concourse process for each pipeline, but ultimately that grew into multiple pipelines configured via the API, and from there it grew into multiple teams, etc. etc. etc.

Expanding API flexibility seems to be the trend, and I do sometimes wonder how much simpler some things could be if we enforced one-pipeline-per-Concourse, but I kind of suspect it would just push the complexity into other places. :thinking:

Reply:
> Side tangent: back in the day, Concourse was literally configured with a single pipeline. It wasn't even called a pipeline, it was just a config file. I resisted going any further than that for a while, wishing people would just run a different Concourse process for each pipeline... 
>
> ... I do sometimes wonder how much simpler some things could be if we enforced one-pipeline-per-Concourse, but I kind of suspect it would just push the complexity into other places. 🤔

Sounds like the Unix philosophy to me - write simple programs that do one thing and do it well, and then compose them to solve more complex problems. The law of conservation of complexity, in context, argues that "pushing the complexity into other places" makes it easier to deal with that complexity by enabling the project to take benefit from the effort that went into the tools in those "other places", instead of re-inventing the wheel.

Reply:
Yeah, that was the idea. I don't quite remember what was the final straw that pushed us in this direction, but it may have been as simple as wanting to access all of the pipelines in one UI at one URL and with one authentication mechanism.

I think there's a lot of value in the current approach from a user-level simplicity standpoint, and we can still have simple UNIXy principles underlying Concourse as part of its internal architecture, with the existing multi-tenant platform providing a consistent opinionated pattern for Concourse users.

I think now we're into kind of a deep philosophical discussion that probably doesn't have much resolution, so I'll mark this conversation as resolved so it doesn't dominate the review process. :slightly_smiling_face: Fun chat though.